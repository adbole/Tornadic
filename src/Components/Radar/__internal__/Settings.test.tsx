import { setLocalStorageItem } from "@test-utils";

import { act, fireEvent, render, screen } from "@testing-library/react";

import DEFAULTS from "Hooks/useLocalStorage.config";

import { Settings } from ".";


vi.mock("svgs/widget", () => ({
    Gear: () => <span>Gear</span>,
}));

vi.mock("Components/Radar/__internal__/Opacity", () => ({
    default: () => <span>Opacity</span>,
}));

test("Matches Snapshot", () => {
    const { container } = render(<Settings />);

    expect(container).toMatchSnapshot();
});

test("When the gear is clicked, the settings popup is shown", () => {
    render(<Settings />);

    expect.soft(screen.queryByText("Opacity")?.parentElement).toHaveStyle({ display: "none" });

    act(() => {
        screen.getByText("Gear").click();
    });

    expect.soft(screen.queryByText("Opacity")?.parentElement).toHaveStyle({ display: "initial" });
});

describe("Changing Settings", () => {
    test("Default settings are loaded", () => {
        render(<Settings />);

        expect(localStorage).toHaveLocalItemValue("radarSettings", DEFAULTS.radarSettings);
    });

    test("Changing the color scheme", () => {
        render(<Settings />);

        const select = screen.getByLabelText("Color Scheme");

        act(() => {
            fireEvent.change(select, { target: { value: 8 } });
        });

        expect(localStorage).toHaveLocalItemValue("radarSettings", {
            ...DEFAULTS.radarSettings,
            colorScheme: 8,
        });
    });

    //Toggle Switches
    describe.each([
        ["Smooth Radar", "smoothing"],
        ["Show Snow", "snow"],
    ] as [string, keyof RadarSettings][])(`%s`, (label, prop) => {
        test.each([
            ["Off -> On", false, true],
            ["On -> Off", true, false],
        ] as [string, boolean, boolean][])(`%s`, (_, defaultState, newState) => {
            const defaults: RadarSettings = {
                ...DEFAULTS.radarSettings,
                [prop]: defaultState,
            };

            setLocalStorageItem("radarSettings", defaults);
            render(<Settings />);

            expect.soft(localStorage).toHaveLocalItemValue("radarSettings", defaults);

            const toggle = screen.getByLabelText<HTMLInputElement>(label);
            expect.soft(toggle.checked).toBe(defaultState);

            act(() => {
                toggle.click();
            });

            expect.soft(toggle.checked).toBe(newState);

            expect.soft(localStorage).toHaveLocalItemValue("radarSettings", {
                ...defaults,
                [prop]: newState,
            });
        });
    });
});
