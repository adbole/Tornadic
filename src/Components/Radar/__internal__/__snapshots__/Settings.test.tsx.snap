// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Matches Snapshot 1`] = `
.emotion-0 {
  -webkit-transition: filter 0.3s ease;
  transition: filter 0.3s ease;
  border: none;
  border-radius: var(--input-border-radius);
  font-size: inherit;
  box-shadow: 0px 4px 10px rgba(0, 0, 0, 0.25);
  margin: 5px 0px;
  color: inherit;
  background-color: transparent;
  box-shadow: none;
  padding: 0;
}

.emotion-0:hover,
.emotion-0:focus {
  cursor: pointer;
}

.emotion-0:active {
  -webkit-filter: brightness(80%);
  filter: brightness(80%);
}

.emotion-0:disabled {
  cursor: not-allowed;
  -webkit-filter: grayscale(0.5);
  filter: grayscale(0.5);
}

.emotion-2 {
  position: absolute;
  right: 40px;
  padding: 10px;
  margin: 10px;
  z-index: var(--z-layer-2);
  width: 200px;
}

.emotion-4 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  justify-content: space-between;
  margin-bottom: 10px;
}

.emotion-8 {
  background-color: var(--background-layer);
  color: inherit;
  border-radius: var(--input-border-radius);
  border: none;
  padding: 5px;
  width: 100%;
}

.emotion-12 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  gap: 10px;
  margin: 5px 0px;
}

.emotion-14 {
  -webkit-appearance: none;
  -moz-appearance: none;
  -ms-appearance: none;
  appearance: none;
  position: relative;
  width: 4rem;
  height: 2rem;
  border-radius: var(--border-radius);
  background-color: var(--background-layer);
  -webkit-transition: background-color 0.3s ease;
  transition: background-color 0.3s ease;
  outline: none;
  cursor: pointer;
}

.emotion-14::after {
  content: "";
  position: absolute;
  top: 50%;
  left: 30%;
  -webkit-transform: translate(-50%, -50%);
  -moz-transform: translate(-50%, -50%);
  -ms-transform: translate(-50%, -50%);
  transform: translate(-50%, -50%);
  border-radius: 50%;
  height: 1.25rem;
  width: 1.25rem;
  background-color: white;
  -webkit-transition: left 0.3s ease;
  transition: left 0.3s ease;
}

.emotion-14:checked {
  background-color: var(--primary);
}

.emotion-14:checked::after {
  left: 70%;
}

<div>
  <button
    class="leaflet-custom-control leaflet-control emotion-0 emotion-1"
    title="Toggle Radar Setttings Visibility"
    type="button"
  >
    <span>
      Gear
    </span>
  </button>
  <div
    class="leaflet-control emotion-2 emotion-3"
    style="display: none;"
  >
    <div
      class="emotion-4 emotion-5"
    >
      <h1
        style="font-weight: 500;"
      >
        Radar Settings
      </h1>
      <button
        class="emotion-0 emotion-1"
        style="margin: 0px;"
        type="button"
      >
        âœ•
      </button>
    </div>
    <label>
      <h2>
        Map Theme
      </h2>
      <select
        class="emotion-8 emotion-9"
        style="font-size: 1.05rem;"
      >
        <option
          value="system"
        >
          System
        </option>
        <option
          value="light"
        >
          Light
        </option>
        <option
          value="light-grey"
        >
          Light Grey
        </option>
        <option
          value="dark"
        >
          Dark
        </option>
        <option
          value="dark-grey"
        >
          Dark Grey
        </option>
      </select>
    </label>
    <label>
      <h2>
        Color Scheme
      </h2>
      <select
        class="emotion-8 emotion-9"
        style="font-size: 1.05rem;"
      >
        <option
          value="0"
        >
          Black and White
        </option>
        <option
          value="1"
        >
          Original
        </option>
        <option
          value="2"
        >
          Universal Blue
        </option>
        <option
          value="3"
        >
          TITAN
        </option>
        <option
          value="4"
        >
          The Weather Channel
        </option>
        <option
          value="5"
        >
          Meteored
        </option>
        <option
          value="6"
        >
          NEXRAD Level III
        </option>
        <option
          value="7"
        >
          Rainbow @ SELEX-SI
        </option>
        <option
          value="8"
        >
          Dark Sky
        </option>
      </select>
    </label>
    <label
      class="emotion-12 emotion-13"
      title="Display all active National Weather Service Alerts within the radar. Alert widget will only display alerts relevant to your location"
    >
      <input
        class="emotion-14 emotion-15"
        type="checkbox"
      />
      <span>
        Alert Mode
      </span>
    </label>
    <label
      class="emotion-12 emotion-13"
      title="Smooths the radar imagry removing pixelation."
    >
      <input
        checked=""
        class="emotion-14 emotion-15"
        type="checkbox"
      />
      <span>
        Smooth Radar
      </span>
    </label>
    <label
      class="emotion-12 emotion-13"
      title="Displays snow on the radar in a different color."
    >
      <input
        class="emotion-14 emotion-15"
        type="checkbox"
      />
      <span>
        Show Snow
      </span>
    </label>
    <span>
      Opacity
    </span>
  </div>
</div>
`;
